sdks/python/apache_beam/runners/dataflow/dataflow_exercise_metrics_pipeline.py

```diff
@@ -0,0 +1,197 @@
+      )
+  ]
+
+  pcoll_names = [
+      'GroupByKey/Reify-out0',
+      'GroupByKey/Read-out0',
+      'map_to_common_key-out0',
+      'GroupByKey/GroupByWindow-out0',
+      'GroupByKey/Read-out0',
+      'GroupByKey/Reify-out0'
+  ]
+  for name in pcoll_names:
+    matchers.extend([
+        MetricResultMatcher(
+            name='ElementCount',
+            labels={
+                'output_user_name': name,
+                'original_name': '%s-ElementCount' % name
+            },
+            attempted=greater_than(0),
+            committed=greater_than(0)
+        ),
+        MetricResultMatcher(
+            name='MeanByteCount',
+            labels={
+                'output_user_name': name,
+                'original_name': '%s-MeanByteCount' % name
+            },
+            attempted=greater_than(0),
+            committed=greater_than(0)
+        ),
+    ])
+  return matchers
+
+
+def fn_api_metric_matchers():
+  """MetricResult matchers with adjusted step names for the FN API DF test."""
+  matchers = common_metric_matchers()
+  return matchers
+
+
+def legacy_metric_matchers():
+  """MetricResult matchers with adjusted step names for the legacy DF test."""
+  # TODO(ajamato): Move these to the common_metric_matchers once implemented
+  # in the FN API.
+  matchers = common_metric_matchers()
+  matchers.extend([
+      # User distribution metric, legacy DF only.
+      MetricResultMatcher(
+          name='distribution_values',
+          namespace=METRIC_NAMESPACE,
+          step='metrics',
+          attempted=DistributionMatcher(
+              sum_value=sum(INPUT),
+              count_value=len(INPUT),
+              min_value=min(INPUT),
+              max_value=max(INPUT)
+          ),
+          committed=DistributionMatcher(
+              sum_value=sum(INPUT),
+              count_value=len(INPUT),
+              min_value=min(INPUT),
+              max_value=max(INPUT)
+          ),
+      ),
+      # Element count and MeanByteCount for a User ParDo.
+      MetricResultMatcher(
+          name='ElementCount',
+          labels={
+              'output_user_name': 'metrics-out0',
+              'original_name': 'metrics-out0-ElementCount'
+          },
+          attempted=greater_than(0),
+          committed=greater_than(0)
+      ),
+      MetricResultMatcher(
+          name='MeanByteCount',
+          labels={
+              'output_user_name': 'metrics-out0',
+              'original_name': 'metrics-out0-MeanByteCount'
+          },
+          attempted=greater_than(0),
+          committed=greater_than(0)
+      ),
+  ])
+  return matchers
+
+
+class UserMetricsDoFn(beam.DoFn):
+  """Parse each line of input text into words."""
+
+  def __init__(self):
+    self.total_metric = Metrics.counter(self.__class__, 'total_values')
+    self.dist_metric = Metrics.distribution(
+        self.__class__, 'distribution_values')
+    self.latest_metric = Metrics.gauge(self.__class__, 'latest_value')
```

Comment:
Maybe add a TODO to add a verifier for this counter once it's supported by other runners? : ) It may be a long-lived TODO, but it's good to have it.

Reply:
done